### PR TITLE
Add indirect-first-instance feature.

### PR DESCRIPTION
**Pull request by mrshannon**
Combining indirect drawing with render bundles is a powerful way to shift work to the GPU.  However, with `firstInstance` locked to 0 the bundle must use dynamic uniform/storage buffer offsets which cannot be changed once the bundle is encoded.  With non-zero values of `firstInstance` the `instance_index` can be used to index into uniform/storage arrays inside the shader, allowing the GPU to change what is rendered each frame, greatly adding to the flexibility of render bundles.

This PR addresses adding an `indirect-first-instance` feature which aims to relax the requirement on `firstInstance = 0` for `drawIndirect` and `drawIndexedIndirect` on hardware which supports `firstInstance > 0`.

NOTE: The decision to split this feature off from https://github.com/gpuweb/gpuweb/pull/1949 was made during the [meeting on 2021-08-02](https://docs.google.com/document/d/1PTyLomUvNd8TyL2a44ZsbjmAprsYM9JE4WYEw0xurH0)

NOTE: This was removed from core (https://github.com/gpuweb/gpuweb/issues/1878) due to the 36% of Android devices which do not support `firstInstance > 0`.

The required backend capabilities to implement `indirect-first-instance` are available on:
* 99% of Vulkan capable desktop GPUs
* 64% of Vulkan capable Android devices
* All DX12 devices
* All Metal devices

In particular, the required capabilities are core to DX12 and Metal, but requires the `drawIndirectFirstInstance` feature on Vulkan.  The corresponding parameter for each backend API are given in the table below:

| Vulkan | DX12 | Metal |
| ------ | ---- | ----- |
| `firstInstance` | `StartInstanceLocation` | `baseInstance` |

***
<a href="https://pr-preview.s3.amazonaws.com/mrshannon/gpuweb/pull/2022.html" title="Last updated on Nov 9, 2021, 9:16 AM UTC (6fb73a5)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/gpuweb/gpuweb/2022/25ce630...mrshannon:6fb73a5.html" title="Last updated on Nov 9, 2021, 9:16 AM UTC (6fb73a5)">Diff</a>

### CODE REVIEW

**Review by kvark — APPROVED** [review]

File: spec/index.bs
Comment (jdashg) [inline spec/index.bs:9531]: s/firstInstnace/firstInstance/

Comment (kainino0x) [reply spec/index.bs:9531]:
```suggestion
Removes the zero value restriction on `firstInstance` in [=indirect draw parameters=] and [=indirect drawIndexed parameters=].
```

**Comment by kainino0x** [comment]
@tidoust does @mrshannon really need to be in the **working** group to land this change? I thought our model was to do IP rubber-stamping in bulk when the spec is imported from the community group to the working group.

**Review by kainino0x — APPROVED** [review]

**Comment by kainino0x** [comment]
For the record: in the meeting we resolved to accept this change.

**Comment by tidoust** [comment]
> @tidoust ipr is blocking this change - does @mrshannon really need to be in the working group to land this change? I thought our model was to do IP rubber-stamping in bulk when the spec is imported from the community group to the working group.

The notion of "rubber-stamping" tends to irk legal departments... The goal of the IPR bot is to alert the Working Group about situations where it is effectively going to incorporate a substantive contribution from a participant who did not make the exact same IPR commitments as WG participants.

CG participants contributions are covered by the [CLA](https://www.w3.org/community/about/process/cla/). WG participants contributions are covered by the [W3C Patent Policy](https://www.w3.org/Consortium/Patent-Policy/). The CLA does provide some level of RF guarantees. The W3C Patent Policy is a bit stronger than the CLA. The goal is to secure as many RF guarantees as we can for the final W3C Recommendation.

To solve this:
- Ideally, @mrshannon, you would be able to make a non-WG-participant commitment by signing the [Final Specification Agreement](https://www.w3.org/community/about/process/final/), which you should be able to do starting from https://labs.w3.org/repo-manager/pr/id/gpuweb/gpuweb/2022. The licensing terms of the FSA are essentially equivalent to those in the W3C Patent Policy.
- Alternatively, @kainino0x, given the CG/WG setting, if the WG group is comfortable with the update, you may override the IPR bot and merge the PR as-is.

**Comment by mrshannon** [comment]
> * Ideally, @mrshannon, you would be able to make a non-WG-participant commitment by signing the [Final Specification Agreement](https://www.w3.org/community/about/process/final/), which you should be able to do starting from https://labs.w3.org/repo-manager/pr/id/gpuweb/gpuweb/2022. The licensing terms of the FSA are essentially equivalent to those in the W3C Patent Policy.

The only option that page currently lists is to _join the GPU for the Web Working Group_. It is very likely I can sign the FSA, but there is currently no means to continue along that route.  I have already signed the CLA, when I joined the CG last year.

**Comment by mrshannon** [comment]
I have permission to sign the FSA, but no way to sign it as the IPR is still blocked on WG only.

**Comment by kainino0x** [comment]
@tidoust can you confirm if there should be a way to sign the FSA without joining the group?

Otherwise we can probably figure out how to get him into the WG as an invited expert? Or we can override the change but we'd want to get WG approval which we don't have a process for.

**Comment by Kangz** [comment]
@tidoust how does one override the IPR bot? I see @ash-nazg mentioned often but I don't know who owns that github account.

@mrshannon has been blocked on this process for a fairly trivial contribution (adding an optional feature that lifts a restriction in the spec). How can he sign the FSA without joining the WG itself? It seems all that matters is the FSA.

**Comment by jdashg** [comment]
FWIW we can always admin-override to merge it, once we know the FSA is signed.

**Comment by munrocket** [comment]
Someone not from WG signed FSA? If yes, how they do that?

**Comment by mrshannon** [comment]
I hate to bump this again, but is there any update on being able to sign the FSA as a CG member?

**Comment by Kangz** [comment]
@tidoust this PR has been blocked for months on CG members not being able to sign the FSA. Any way we can escalate?

**Comment by tidoust** [comment]
Oh no! Sorry this has languished for months, mentions of my name did not trigger the notifications I was expecting so I simply hadn't seen this.

There is a way to ask for non-member commitments through the IPR bot with two caveats:
1. I, or someone else from the team, need to push the right button. So you were right to ping me and I'm very sorry that I missed these repeated pings :/
2. That is per PR.

 I just did that for this PR. @mrshannon, please check your email inbox for a link to the relevant online form.

**Comment by mrshannon** [comment]
Agreement completed. Once this PR is merged I will rebase #1949 on this and ping that one as well. Thanks.

**Comment by tidoust** [comment]
And IPR bot turned green as a result, so all good!

**Pull request merged**
